### PR TITLE
Create game rules to check if a game is won or tied

### PR DESCRIPTION
**Pull request by agonzalez0515**
This PR implements looping until the game is over. It checks if the game is won or tied after each move.

### CODE REVIEW

File: spec/ttt/game_rules_spec.clj [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                                  5   6   7])))
+    (it "checks if the game is won for a row of O's"
+      (should= true (rules/win? ["X"  1  "X"
+                                 "O" "O" "O"
+                                  5   6   7])))
+    (it "checks if the game is won for a column of X's"
+      (should= true (rules/win? ["X" "O" "O"
+                                 "X"  4  "O"
+                                 "X"  6   7])))
+    (it "checks if the game is won for a column of O's"
```
Comment (tbaik) [inline spec/ttt/game_rules_spec.clj]: I think rather than having 2 tests for the column of different markers, we should test for each of the three columns. I'm not sure if everyone does this but I like it mix up the parameters like you did with the O's and X's in similar tests. So, I think you can just keep these two tests and write one more for the 3rd column! (same thing for row)

File: spec/ttt/player_spec.clj [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
+            [ttt.player :as player]))
 
 (describe "Player"
   (context "#players"
     (it "creates a player with marker X"
-      (should= "X" (:marker (first players))))
+      (should= "X" (:marker (first player/players))))
     (it "creates a player with marker O"
-      (should= "O" (:marker (second players)))))
+      (should= "O" (:marker (second player/players)))))
```
Comment (tbaik) [inline spec/ttt/player_spec.clj]: I think instead of creating two tests for this, we can combine them into one and assert on whatever `player/players` returns.

File: src/ttt/core.clj [hunk 3]
```diff
@@ -14,18 +15,21 @@
         (fn [current-state]
            (merge current-state {:board board :players players :current-player (first players)}))))
 
-(defn play 
+(defn play
```
Comment (tbaik) [inline src/ttt/core.clj:16]: Let's write tests for this. Now, I know this is probably the most difficult part... Would you like to pair with one of the mentors or someone else from the office on this one to get you started?

File: src/ttt/core.clj [hunk 4]
```diff
@@ -14,18 +15,21 @@
...
   []
-  (ui/print-board (:board @game-state))
-  (set-game-state 
-   (board/make-move (:board @game-state) (player/get-move) (:marker (:current-player @game-state))) 
-   (player/switch-players (:players @game-state)))
-  (ui/print-board (:board @game-state)))
+  (loop [state @game-state]
```
Comment (tbaik) [inline src/ttt/core.clj]: Now that you have a loop, do you think you still need the game state atom?

File: src/ttt/core.clj [hunk 5]
```diff
@@ -14,18 +15,21 @@
...
+    (ui/print-board (:board state))
+  (if (rules/game-over? (:board state))
+    (println "game over")
```
Comment (tbaik) [inline src/ttt/core.clj]: seems like this should go in the ui namespace

File: src/ttt/game_rules.clj [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  [board]
+  [(create-rows board) (create-columns board) [(create-left-to-right-diagonal board) (create-right-to-left-diagonal board)]])
+
+(defn- check-group-of-cells
+  [cells]
+  (map (fn[collection] (= 1 (count (set collection)))) cells))
+
+(defn- available-spaces?
+  [board]
+  (boolean (some true? (map number? board))))
```
Comment (tbaik) [inline src/ttt/game_rules.clj]: I believe you can just do (some number? board) for the same thing. Also, in other cases, I believe you can use `some?` instead of doing `some true?`

Comment (tbaik) [reply src/ttt/game_rules.clj]: one more good thing is that it seems like `some?` returns a boolean false for the falsey case

File: src/ttt/game_rules.clj [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+(defn win?
+  [board]
+  (boolean (some true? (flatten (map check-group-of-cells (create-groups-of-cells board))))))
+
+(defn tie?
+  [board]
+  (and (not (available-spaces? board)) (not (win? board))))
```
Comment (tbaik) [inline src/ttt/game_rules.clj]: I personally don't care but some people are pretty particular about not having a lot of negatives in their conditional statements. So the alternative would be `(not (or (available-spaces? board) (win? board)))`. You should pick whichever way seems easiest to understand!

**Review by tbaik — CHANGES_REQUESTED** [review]

File: src/ttt/game_rules.clj (hunk 6 above)
Comment (tbaik) [reply src/ttt/game_rules.clj]: acttually sorry, I tried it on repl and it's not what I thought. It's just `not-nil`

File: spec/ttt/game_rules_spec.clj [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+      (should= true (rules/tie? ["O" "X" "O"
+                                 "O" "X" "X"
+                                 "X" "O" "X"])))
+    (it "returns false if the game is not tied"
+      (should= false (rules/tie? ["O" 1 "O"
+                                  "O" "X" "X"
+                                  "X" "O" "X"]))))
+  (context "#game-over?" 
+    (it "returns true if the game is over"
+      (should= true (rules/win? ["O" "X" "X"
```
Comment (kkrull) [inline spec/ttt/game_rules_spec.clj]: Did you mean to call `rules/game-over?`

File: spec/ttt/game_rules_spec.clj [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                                  5  "O"  7])))
+    (it "checks if the game is won for a diagonal of X's"
+      (should= true (rules/win? ["O" "X" "X"
+                                 "O" "X" "O"
+                                 "X"  6   7])))
+    (it "checks if the game is won for a diagonal of O's"
+      (should= true (rules/win? ["O" "X" "X"
+                                  3  "O" "O"
+                                 "X"  6  "O"]))))
+  (context "#tie?"
```
Comment (kkrull) [inline spec/ttt/game_rules_spec.clj:48]: It looks like you may have missed a case for `rules/tie?`: when there aren't any available spaces, and there is a winner.  If I'm not mistaken, that means this incorrect version of `rules/tie?` might still pass the tests here

```clojure
(defn rules/tie?
  [board]
  (not (available-spaces? board)))
```

This in turn makes me wonder: were these tests written after you already implemented `rules/tie?`?

File: spec/ttt/game_rules_spec.clj [hunk 10]
```diff
@@ -0,0 +1,53 @@
+(ns ttt.game-rules-spec
+  (:require [speclj.core :refer [describe context it should=]]
+            [ttt.game-rules :as rules]))
+
+
+(describe "Game Rules"
+  (context "#win?"
```
Comment (kkrull) [inline spec/ttt/game_rules_spec.clj:11]: _food for thought; no action required_

The discussion topic Tony started is among my favorites: what is the minimum set of tests required that:

* fully _exercises_ the behavior of the production code (i.e. you drove every expression in the production code until you couldn't write any more tests)
* fully _describes_ the behavior of the production code (i.e. the text in the `it` clauses reads well to human beings, if you were to remove all the test code)

I like to think about this as if my tests are teaching the production code to do progressively more difficult things, like:

* always return false (given a game board with no winner)
* learn to look for a specific player X in a specific row (the first row)
* learn to iterate over all rows by checking for a winner in the 2nd row
* learn about another player, O, by looking for O to win in any row (it already knows how to check rows, at this point)
* learn to check the first column, then all columns, by checking for a win in the first and second columns
* learn to check both diagonals, by checking for a win in each diagonal
* if necessary, make sure a row, column, or diagonal that is occupied by multiple players returns false (depends on whether the production code was able to get away with a lesser check, up to this point)

There are a number of ways to go about adding/modifying tests to go through enough steps to make your production code do everything it needs to do.  I think it's kind of neat to think about what is the minimal set of tests that still meets the exercise/describe criteria above, but I wouldn't get too carried away looking for a perfect solution.

The bottom line: Keep writing tests until the production code does everything it should, and try to keep them organized in some fashion that seems logical to humans, as you go.  If you wind up with a few extra tests (within reason), it's no big deal.

File: spec/ttt/player_spec.clj (hunk 2 above)
Comment (kkrull) [reply spec/ttt/player_spec.clj]: This is a good example of "one _logical_ assertion, per test".  You could think of saying "it creates enough players to play the game" as the behavior that's meaningful to humans and the two `should=` checks inside it.  The two assertions would still be written–and the production code driven–by separate red-green-refactor cycles.

File: spec/ttt/ui_spec.clj [hunk 11]
```diff
@@ -1,15 +1,15 @@
...
-      (with-out-str (print-start-game-message [{:marker "X"} {:marker "O"}]))))
+    (should= "Let's play tic-tac-toe.\nPlayer 1 is X\nPlayer 2 is O\nEnter a number between 0-8 to select a spot.\n" 
+      (with-out-str (ui/print-start-game-message [{:marker "X"} {:marker "O"}]))))
   (it "prints a board"
     (should= "\n0 | 1 | 2\n---------\n3 | 4 | 5\n---------\n6 | 7 | 8\n" 
-      (with-out-str (print-board [0 1 2 3 4 5 6 7 8]))))
+      (with-out-str (ui/print-board [0 1 2 3 4 5 6 7 8]))))
   (it "prints an invalid move message"
-      (should= "Invalid move. Please try again.\n" (with-out-str(print-invalid-move-message)))))
+      (should= "Invalid move. Please try again.\n" (with-out-str(ui/print-invalid-move-message)))))
```
Comment (kkrull) [inline spec/ttt/ui_spec.clj]: Looks like there's some whitespace missing just before `(ui/print-invalid-move-message)`.  This seems to work as far as I can see from the style guide, but the whitespace aids in readability.

File: src/ttt/core.clj (hunk 3 above)
Comment (kkrull) [reply src/ttt/core.clj:16]: Agreed - I think this will be a helpful exercise in practicing TDD on things that people often write off as "impossible" (and hence, where the bugs go to hide).  Exposing dependencies, testing with some sort of test double, and driving the algorithm incrementally is often the trick here.  (Sometimes it seems really hard in the beginning, but it winds up not being that hard once you get past the first couple of tests).

File: src/ttt/game_rules.clj [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      (into [] (map get-cell-value row))))
+
+(defn- create-left-to-right-diagonal
+  [board]
+  (let [rows (create-rows board)]
+    (create-diagonal rows)))
+
+(defn- create-right-to-left-diagonal
+  [board]
+  (let [rows (reverse (create-rows board))]
```
Comment (kkrull) [inline src/ttt/game_rules.clj]: It looks like your code has to work pretty hard to re-structure a flattened list/vector into rows, columns, and diagonals.  On the bright side, you do have all this code pretty well isolated into its own namespace such that nobody else really has to know what the board looks like inside.  Would another method of organizing the game board's spaces (internally) make these functions easier to write?

File: src/ttt/game_rules.clj [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+(defn- get-board-size
+  [board]
+  (math/sqrt (count board)))
+
+(defn- create-columns
+  [board]
+ (apply mapv vector (partition (get-board-size board) board)))
+
+(defn- create-rows
```
Comment (kkrull) [inline src/ttt/game_rules.clj]: This might be good to call when printing the game board.  If so, this would have to be promoted to a public method and test(-driven).

**Review by kkrull — COMMENTED** [review]
Overall it looks like you're getting more comfortable with Clojure, and separating concerns into separate namespaces seems to be coming more naturally.  It looks like you're doing a good job of testing your code; the places where there may be some gaps are the places that require different testing techniques.  All that is to say: I'm happy to see you applying everything you've learned so far, to this new code.

There don't seem to be too many pieces of non-board code that need to know what the board looks like inside, but there may still be some places in the UI code that know a bit more than necessary.  Encapsulation is just as valuable and possible in functional languages, even thought it's not emphasized as much.  Hiding the board's internal data here will help reduce coupling and brittleness, just like it did in Ruby.  That may also have a positive ripple effect of making the UI code, the board row/column/diagonal query functions, and the win/draw code all easier to test.

As usual there's plenty to talk and think about, but hopefully you'll be able to notice an elevation in the kind of dialogue we're having.  Comments are less like "this is a bug" and more about philosophy and maintainability.  Keep up the good work!

File: spec/ttt/game_rules_spec.clj (hunk 8 above)
Comment (agonzalez0515) [reply spec/ttt/game_rules_spec.clj]: oops yes!

File: spec/ttt/game_rules_spec.clj (hunk 9 above)
Comment (agonzalez0515) [reply spec/ttt/game_rules_spec.clj:48]: I think I wrote the tests first? But there is a chance i didnt.  Probable reason why it passed was because i didn't write the most minimal code to make it pass. 🙈 Need to get better at not getting ahead of the test. 

Now I'm thinking if I should be checking `(not (win? board))` in the `tie?` function at all. My `game-over?` checks for a win first, and then for a tie (which also checks for no win). Am I being redundant?

File: src/ttt/core.clj (hunk 3 above)
Comment (agonzalez0515) [reply src/ttt/core.clj:16]: I think I'm going to try and redo this function test driving it now that I have a better sense of how `recur` works. That was the part that I was stuck on for awhile.

Comment (kkrull) [reply src/ttt/core.clj:16]: Cool.  That's a perfectly viable approach, in my view (I do it all the time, so it must be ok :trollface: )

Pro tip: Keep the old function around and look at it for reference, while you're testing-driving the new version.  You can always test-drive `play-new` and then get rid of the old version when you're done writing the new one.  (maintaining the old function under the old name, until the new one is ready, is a good way to keep all those test passing).

File: spec/ttt/game_rules_spec.clj (hunk 9 above)
Comment (kkrull) [reply spec/ttt/game_rules_spec.clj:48]: That does happen from time to time: you thought you were taking the smallest possible leap in the production code that got your tests to pass, only to realize later you could have taken a smaller step and added an additional test.  So it's all good; seeing gaps is a good way to make a reasonable guess about the process that was used to create the code.

You can implement `win?` and `tie?` by having one use the other for (part of) its algorithm.  Either way is fine in my view; just make sure you don't have them pointing to each other (you'll get an endless loop).

File: src/ttt/core.clj (hunk 4 above)
Comment (agonzalez0515) [reply src/ttt/core.clj]: doesn't look like it! I refactored it to take a starting game state at the beginning when `play` is called, and i just create a new state each time it finishes the loop.

File: spec/ttt/core_spec.clj [hunk 14]
```diff
@@ -1,14 +1,59 @@
...
-  (:require [speclj.core :refer :all]
-            [ttt.core :refer :all]))
+  (:require [speclj.core :refer [describe context it should= should-invoke stub with-stubs]]
+            [ttt.core :as game]
+            [ttt.game-rules :as rules]
+            [ttt.ui :as ui]
+            [ttt.board :as board]
+            [ttt.player :as player]))
 
+(def current-state {:board [0 1 2 3 4 5 6 7 8]
```
Comment (tbaik) [inline spec/ttt/core_spec.clj]: should this be initial state since current state implies it changes as the game is being played?

File: spec/ttt/core_spec.clj [hunk 15]
```diff
@@ -1,14 +1,59 @@
...
+(def players [{:marker "O"} {:marker "X"}])
 
 
+(describe "Game"          
+  (context "#play"
+    (with-stubs)
+    
+    (it "calls game over message when the game is over"
+      (with-redefs [rules/game-over? (stub :rules/game-over? {:return true})]
+
```
Comment (tbaik) [inline spec/ttt/core_spec.clj]: I think for me personally, the whitespace here makes it a bit confusing because I expected the `it` block to not have a whitespace in the middle of it. I do see why you'd do that though. Is there a guideline for this somewhere?

File: spec/ttt/core_spec.clj [hunk 16]
```diff
@@ -1,14 +1,59 @@
...
+    
+    (it "prints new boards after each player makes a move"
+      (with-redefs [player/get-move (let [results (atom [4 8])]
+                                      (fn [] (ffirst (swap-vals! results rest))))
+                    board/make-move (let [results (atom [[0 1 2 3 "X" 5 6 7 8] [0 1 2 3 "X" 5 6 7 "O"]])]
+                                      (fn [_a _b _c] (ffirst (swap-vals! results rest))))
+                    rules/game-over? (let [results (atom [false false true])]
+                                        (fn [_] (ffirst (swap-vals! results rest))))]
+
+      (should= "\n0 | 1 | 2\n---------\n3 | 4 | 5\n---------\n6 | 7 | 8\n\n0 | 1 | 2\n---------\n3 | X | 5\n---------\n6 | 7 | 8\n\n0 | 1 | 2\n---------\n3 | X | 5\n---------\n6 | 7 | O\nGame over\n"
```
Comment (tbaik) [inline spec/ttt/core_spec.clj]: It may be easier for people to read this if you either put this into a variable with a descriptive name or if you separate it out!

File: src/ttt/core.clj [hunk 17]
```diff
@@ -2,30 +2,33 @@
...
+  (assoc state :board board :players players :current-player (first players)))
 
+(defn play
+  [starting-game-state]
+  (loop [state starting-game-state]
+    (ui/print-board (:board state))
+  (if (rules/game-over? (:board state))
+    (ui/print-game-over-message)
+    (let [next-board (board/make-move 
+                      (:board state)
```
Comment (tbaik) [inline src/ttt/core.clj]: I think 23,24,25 need one more spaces each, it doesn't look like a tab to me 😁

**Comment by tbaik** [comment]
Other than just the few comments I left just now, it looks really good! Please feel free to merge once you've made those changes!

File: spec/ttt/core_spec.clj [hunk 18]
```diff
@@ -14,34 +14,46 @@
...
         
       (should= "\n0 | 1 | 2\n---------\n3 | 4 | 5\n---------\n6 | 7 | 8\n\n0 | X | 2\n---------\n3 | 4 | 5\n---------\n6 | 7 | 8\nGame over\n" 
-               (with-out-str (game/play current-state)))))))
+               (with-out-str (game/play current-state)))))
+    
+    (it "prints new boards after each player makes a move"
+      (with-redefs [player/get-move (let [results (atom [4 8])]
+                                      (fn [] (ffirst (swap-vals! results rest))))
+                    board/make-move (let [results (atom [[0 1 2 3 "X" 5 6 7 8] [0 1 2 3 "X" 5 6 7 "O"]])]
+                                      (fn [_a _b _c] (ffirst (swap-vals! results rest))))
```
Comment (kkrull) [inline spec/ttt/core_spec.clj]: This is a good example of a convention often used in functional programming: putting `_` in front of a variable you don't intend to use.  Also a good example of how `!` typically indicates a function that mutates state or that has a side-effect.

Nothing actionable here - just reinforcing what has probably already come up in your learning over the past couple of weeks.  Nice work on the new tests BTW! 👍

**Review by kkrull — APPROVED** [review]
